### ISSUE TITLE
[wasm] Does `browser-wasm` officially support web worker environment?

### ISSUE THREAD

**Issue by Elringus** [dotnet/runtime#80045]
After reading https://devblogs.microsoft.com/dotnet/use-net-7-from-any-javascript-app-in-net-7/ and checking the samples I had an impression the intended use case is to run it on the main browser thread only, but it seems to be working fine under worker threads as well (#77946).

Is this something just happen to work for now, but can break in the future, or worker environment is officially supported and will be maintained under the `browser-wasm` runtime (ie, wasm module and associated JS wrapper won't use `window` or other APIs not available for workers)?

**Comment by am11** [comment]
Glad to see dotnet wasm working in constrained / sandbox environments as well! 8-)

In .NET 7, dotnet.js and dotnet.wasm were modularized and support for non-browser environments was added (e.g. node.js, v8 and [for deno](https://gist.github.com/am11/159ad22a8f5131f3d8811f1c8b34d75b)). Prior to .NET 7, there were only a handful of DOM / `window` usages in the core implementation, which were refactored in .NET 7 with guarded conditions like `if (typeof window.foo === 'function') window.foo; /* else useSomethingElse() */` as part of the modularization work.

I think (hope) moving forward, we won't regress it by (re)introducing unconditional dependencies on DOM or `window` and therefore keep supporting sandbox environment such as Web Workers. Perhaps we can add some tests to preserve this desired state.

**Comment by Elringus** [comment]
> I think (hope) moving forward, we won't regress it by (re)introducing unconditional dependencies on DOM or window and therefore keep supporting sandbox environment such as Web Workers.

That's exactly what I'm concerned about, especially with all the docs and samples only mention running either on main browser thread or under node. Would be great to have an official statement somewhere (eg, https://learn.microsoft.com/en-us/aspnet/core/client-side/dotnet-interop) and associated tests covering workers (or constrained environments in general).

I'm currently using a patched .NET 6 fork to make it compatible with both node and worker environments. The prospect of migrating to stock runtime is exciting, but I need to be sure it won't suddenly break at some point.

**Comment by Kuinox** [comment]
> but it seems to be working fine under worker threads as well (https://github.com/dotnet/runtime/issues/77946).

Hi, I opened this issue, yes, currently I run successfuly [a .NET compiler toolchain](https://playground.draco-lang.org/) in a webworker.  
Out of the box, it works on chrome, but not on firefox, because  
[Firefox doesn't support ES modules in webworkers yet](https://bugzilla.mozilla.org/show_bug.cgi?id=1247687), and [.NET dropped support for cjs modules](https://github.com/dotnet/runtime/issues/72990).
To work around this, I use vite to convert `dotnet.js` into CJS:  
```js
await viteBuild(defineConfig({
    build: {
        lib: {
            entry: path.resolve(binFolder, 'dotnet.js'),
            name: 'dotnet',
            fileName: 'dotnet',
            formats: ['umd']
        },
        rollupOptions: {
            external: ['dotnet.wasm'],
            output: {
                esModule: false,
                format: 'cjs',
                compact: true
            }
        },
        outDir: outDir
    }
}));
```

Now in my webworker I import the modified dotnet.js, and an odd startup: 
```ts
importScripts('./dotnet.js');
async function main() {
    const monoCfg = await firstMessagePromise;
    const dotnet = self.dotnet.dotnet;
    dotnet.moduleConfig.configSrc = null;
    const { setModuleImports, getAssemblyExports, } = await dotnet
        .withConfig(
            monoCfg
        )
        .create();
    await dotnet.run();
}
main();
```
And voilà.

**Comment by Elringus** [comment]
I think an ultimate test for this would be making a sample (similar to the TodoMVC and React mentioned in the article) of a [VS Code web extension](https://code.visualstudio.com/api/extension-guides/web-extensions) built using the new JSInterop and compiled for wasm target. The requirements for web extension are very strict: they have to be bundled as a single JS file, can't use any imports or node/browser APIs, etc. That would also refine various APIs, as, for example, new JSInterop requires loading modules for imported function at runtime from C#, which is not allowed in a VS Code web extension (#87365).

**Comment by Kuinox** [comment]
I ran some experiments for https://playground.draco-lang.org/, I inlined the imports with esbuild. There is still imports, but no network requests.

**Comment by Elringus** [comment]
> I ran some experiments (locally, didnt deployed it) for https://playground.draco-lang.org/, I inlined the imports with esbuild. There is still imports, but no network requests.
Do you have `[JSImport]` methods there? If so, how do you use `JSHost.ImportAsync`? In my tests it didn't work when I've attempted importing the main js file; the application just hang.

**Comment by Kuinox** [comment]
Yes I do use [JSImport] methods, but I don't use JSHost.ImportAsync.

**Comment by Elringus** [comment]
> Yes I do use [JSImport] methods, but I don't use JSHost.ImportAsync.

Ah, I wasn't aware about `setModuleImports`; thought that the only way to import a function from JS is to use `JSHost.ImportAsync`. Thanks for the insight!

Though, I still think VS Code extension would be an ultimate sample for constraint environments, as it's running in a very restricted web worker. Using general browser web worker won't cover all use cases.

**Comment by pavelsavara** [comment]
Running the runtime's main on the web worker is something I didn't try yet. Is this issue about that, should we make it more explicit ?

Work on threads (via web workers) in in progress. 
For extensions, we mostly fixed CSP. 

Regarding bundling all of it into single file, I think external solutions like webpack are way to go, to include also user code. 
For the start it would be good to see if our current code is preventing webpack in some way. And report it here.
https://github.com/dotnet/runtime/issues/86162

**Comment by Elringus** [comment]
> Running the runtime's main on the web worker is something I didn't try yet. Is this issue about that, should we make it more explicit ?

More specifically running it in constraint environments, live VS Code's web extensions; there are other constraint enthronement cases, like the deno runtime or plug-ins for various web apps, but I thought a VS Code extension sample would be more appropriate/useful for the community. There are VS Code extensions written in C# and depending on a standalone .NET runtime installed on the machine (eg, C# LSP); because of that they are not compatible with the new web extension format. Making a sample would help them more confidently migrate to the web format.

**Comment by Kuinox** [comment]
> Running the runtime's main on the web worker is something I didn't try yet.

I can tell you it works.  
I have a demo doing it, [it's availabe here.](playground.draco-lang.org/).  
The source are here: https://github.com/Draco-lang/Compiler/tree/main/src/Draco.Editor.Web/app  
Most of the time was spent into bundling it, and figuring out which API to use.

**Comment by Elringus** [comment]
> Well, LSP and Roslyn are probably better off on CoreCLR, with it's OS level file system access and native JIT performance.

Sure, but it won't work in https://vscode.dev this way. I had an impression vscode is pushing extensions to support web.

**Comment by pavelsavara** [comment]
Re-reading it now, there are few updates
- we now have demo effort for running in web-worker https://github.com/dotnet/runtime/issues/95452
- Firefox has ES6 modules in WebWorker now
- VS code is leaning into WASI https://code.visualstudio.com/blogs/2023/06/05/vscode-wasm-wasi
- making dotnet single-file is still on our radar, but no progress yet https://github.com/dotnet/runtime/issues/86162
- making dotnet friendly to constrained environments (like extension) is this issue

**Comment by elringus** [comment]
https://github.com/dotnet/runtime/issues/91558 can probably be merged into this issue?
